### ISSUE TITLE
tuple element types wrongly mixed as union

### ISSUE THREAD

**Issue by ydirson** [python/mypy#20188]
**Bug Report**

When a variable conditionally gets one or the other of a set of fixed-length tuples, instead of being inferred as a union of said tuple types, `mypy` sees an arbibrary-length tuple whose elements are typed with a union of all element types involved.  This results false-positive errors being reported.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&gist=4780da0a891d33c46ee7ac0b55a05f24

**Expected Behavior**

Similar to what `pyright` reveals, a union of two tuple types with defined length and element types, and no error when assigning :
```
  /home/user/omaha/omaha/uitest/mypy-test.py:9:17 - information: Type of "item" is "tuple[Literal[1], Literal['a'], list[bool]] | tuple[Literal[1], Literal['a']]"
```

**Actual Behavior**

The type is instead inferred to be a tuple of arbitrary length, with all elements allowed to have any of the types seen inside the tuple:
```
main.py:9: note: Revealed type is "builtins.tuple[builtins.int | builtins.str | builtins.list[builtins.bool], ...]"
main.py:10: error: List item 0 has incompatible type "tuple[int | str | list[bool], ...]"; expected "tuple[int, str] | tuple[int, str, Sequence[bool]]"  [list-item]
```

Note that when using just `item = (1, "a")`, removing the conditional part, the tuple is properly identified with fixed types, and the assignment is validated by mypy:
```
main.py:9: note: Revealed type is "tuple[builtins.int, builtins.str]"
```

**Your Environment**

- Mypy version used: 1.18.2
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12, 3.14.0rc2

**Comment by randolf-scholz** [comment]
I believe this is less about join vs union and more about how type checkers deal with tuple concatenation.

This appears to be a hole in the type system: the current spec does not allow multiple unpacked `TypeVarTuple`, so it is impossible to even properly express tuple concatenation.

```python
from typing import reveal_type

def concat_tuples[*Us, *Vs](
    left: tuple[*Us],
    right: tuple[*Vs],
) -> tuple[*Us, *Vs]:  # Illegal according to spec
    return reveal_type( left + right )
```

`pyright` seems to do two things to deal with this case:

1. It allows unions of TypeVarTuple members to convert `tuple[*Us]` to `tuple[Union[*Us], ...]`
2. It special cases `+` for `tuple`. One can see this by checking that `left.__add__(right)` gives a different result.

**Comment by CaselIT** [comment]
I think that the spec could be interpreted as restricting only ambiguous cases, like in the example given https://peps.python.org/pep-0646/#multiple-type-variable-tuples-not-allowed.

The concat_tuples example is not ambiguous. 

pyright seems to work with any object not only tuples:
```py
from typing import assert_type

class Select[*T]:
    pass

class Table[*T]:
    pass

def select[*T1, *T2](t1: Table[*T1], t2: Table[*T2]) -> Select[*T1, *T2]: ...

a: Table[int, float] = Table()
b: Table[str, float] = Table()
assert_type(select(a, b), Select[int, float, str, float])
```
